developer/src/kmc-kmw/src/compiler/write-compiled-keyboard.ts

```diff
@@ -0,0 +1,665 @@
+    }
+  }
+  return '';
+}
+
+function JavaScript_SetupProlog() {
+  if(IsKeyboardVersion10OrLater()) {
+    return 'if(typeof keyman === \'undefined\') {'+nl+
+      FTabStop+'console.log(\'Keyboard requires KeymanWeb 10.0 or later\');'+nl+
+      FTabStop+'if(typeof tavultesoft !== \'undefined\') tavultesoft.keymanweb.util.alert("This keyboard requires KeymanWeb 10.0 or later");'+nl+
+      '} else {';
+  }
+  return '';
+}
+
+function JavaScript_SetupEpilog() {
+  if(IsKeyboardVersion10OrLater()) {
+    return '}';
+  }
+  return '';
+}
+
+function HasSupplementaryPlaneChars() {
+  return false; // TODO
+/*  function StringHasSuppChars(p: PWideChar): Boolean;
+  begin
+    if not Assigned(p) then
+      Exit(False);
+
+    while p^ <> #0 do
+    begin
+      if Char.IsSurrogate(p, 0) then
+        Exit(True);
+      p := incxstr(p);
+    end;
+
+    Result := False;
+  end;
+
+var
+  I: number;
+  fsp: PFILE_STORE;
+  fgp: PFILE_GROUP;
+  j: number;
+  fkp: PFILE_KEY;
+begin
+  fsp := fk.dpStoreArray;
+  for i := 0 to number(fk.cxStoreArray) - 1 do
+  begin
+    if StringHasSuppChars(fsp.dpString) then
+      Exit(True);
+    Inc(fsp);
+  end;
+
+  fgp := fk.dpGroupArray;
+  for i := 0 to number(fk.cxGroupArray) - 1 do
+  begin
+    fkp := fgp.dpKeyArray;
+    for j := 0 to number(fgp.cxKeyArray) - 1 do
+    begin
+      if StringHasSuppChars(fkp.dpContext) or
+         StringHasSuppChars(fkp.dpOutput) then
+        Exit(True);
+      Inc(fkp);
+    end;
+
+    if StringHasSuppChars(fgp.dpMatch) or
+       StringHasSuppChars(fgp.dpNoMatch) then
+      Exit(True);
+
+    Inc(fgp);
+  end;
+
+  Result := False;
+end;*/
+}
+
+export function RuleIsExcludedByPlatform(keyboard: KMX.KEYBOARD, fkp: KMX.KEY): boolean {
+  if(fkp.dpContext == null || fkp.dpContext == '') {
+    return false;
+  }
+
+  let x = 0;
+  while(x < fkp.dpContext.length) {
+    let rec = ExpandSentinel(keyboard, fkp.dpContext, x);
+    if(rec.IsSentinel &&
+        (rec.Code == KMX.KMXFile.CODE_IFSYSTEMSTORE) &&
+        (rec.IfSystemStore.dwSystemID == KMX.KMXFile.TSS_PLATFORM) &&
+        rec.IfSystemStore.Store.dpString.includes('native')) {
+      if(rec.IfSystemStore.Store.dpString.match(/windows|desktop|macosx|linux/)) {
+        return true;
+      }
+    }
+    x = incxstr(fkp.dpContext, x);
+  }
+
+  return false;
+}
+
+function WriteBeginStatement(keyboard: KMX.KEYBOARD, name: string, groupIndex: number): string {
+  const fgp = keyboard.groups[groupIndex];
+  return `${FTabStop}this.${name}=function(t,e) {${nl}`+
+         `${FTabStop+FTabStop}return this.g${JavaScript_Name(groupIndex, fgp.dpName)}(t,e);${nl}`+
+         `${FTabStop}};${nl}`;
+}
+
+/**
+ * Converts a modifier bit mask integer into its component bit flags
+ *
+ * @param FBitMask A KMX modifier bitmask value
+ *
+ * @return string of JavaScript code, e.g. 'modCodes.SHIFT | modCodes.CTRL /* 0x0030 ./'
+**/
+export function FormatModifierAsBitflags(FBitMask: number): string {
```

Comment:
For future reference:

https://github.com/keymanapp/keyman/blob/c61655a51495a5adafc734f611f03d29cff44b6d/common/web/keyboard-processor/src/text/codes.ts#L2-L5

Serves pretty much the same purpose.  Modularization would need to land first, of course.

Just noting areas that could use common definitions eventually once modularization (🧩) lands.